projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,169 @@
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_HI_THRESH,
+                                         ADS1015_HI_THRESH_REGISTER_MSB,
+                                         ADS1015_HI_THRESH_REGISTER_LSB));
+  GPIOSettings gpio_settings = {
+    .direction = GPIO_DIR_IN,  //
+  };
+  InterruptSettings it_settings = {
+    .type = INTERRUPT_TYPE_INTERRUPT,
+    .priority = INTERRUPT_PRIORITY_NORMAL,
+  };
+  status_ok_or_return(gpio_init_pin(ready_pin, &gpio_settings));
+  status_ok_or_return(gpio_it_register_interrupt(ready_pin, &it_settings, INTERRUPT_EDGE_RISING,
+                                                 prv_interrupt_handler, storage));
+  // Mask the interrupt until channels are enabled by the user.
+  return gpio_it_mask_interrupt(ready_pin, true);
+}
+
+// This function enable/disables channels, and registers callbacks for each channel.
+StatusCode ads1015_configure_channel(Ads1015Storage *storage, Ads1015Channel channel, bool enable,
+                                     Ads1015Callback callback, void *context) {
+  if (storage == NULL || channel >= NUM_ADS1015_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  if (storage->channel_enable_bitset == ADS1015_BITSET_EMPTY && enable) {
+    // Activate the interrupt since the first channel is being enabled.
+    status_ok_or_return(gpio_it_mask_interrupt(&storage->ready_pin, false));
+    status_ok_or_return(prv_set_channel(storage, channel));
+  } else if (!enable) {
+    storage->channel_readings[channel] = ADS1015_DISABLED_CHANNEL_READING;
+  }
+
+  prv_mark_channel_enabled(channel, enable, &storage->channel_enable_bitset);
+  if (storage->channel_enable_bitset != ADS1015_BITSET_EMPTY) {
+    storage->current_channel = __builtin_ffs(storage->channel_enable_bitset) - 1;
+  } else {
+    // Mask interrupt if all channels are disabled.
+    status_ok_or_return(gpio_it_mask_interrupt(&storage->ready_pin, true));
+  }
+  storage->channel_enable_bitset_helper = storage->channel_enable_bitset;
+  storage->channel_callback[channel] = callback;
+  storage->callback_context[channel] = context;
+  return STATUS_CODE_OK;
+}
+
+// Reads raw 12 bit conversion results which are expressed in two's complement format.
+StatusCode ads1015_read_raw(Ads1015Storage *storage, Ads1015Channel channel, int16_t *reading) {
+  if (channel >= NUM_ADS1015_CHANNELS || storage == NULL || reading == NULL) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  *reading = storage->channel_readings[channel];
+  return STATUS_CODE_OK;
+}
+
+// Reads conversion value in mVolt.
+StatusCode ads1015_read_converted(Ads1015Storage *storage, Ads1015Channel channel,
+                                  int16_t *reading) {
+  if (channel >= NUM_ADS1015_CHANNELS || storage == NULL || reading == NULL) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  int16_t raw_reading = ADS1015_READ_UNSUCCESSFUL;
```

Comment:
You could probably just make this `INT16_MAX` or something.

Reply:
no macro necessary?
I want it to not be confused with disabled channel reading.

Reply:
What benefit would that have? You don't expose this, so how would the user tell? I think it would make more sense to have an "invalid" result that the user can depend on. Alternatively, wouldn't it make more sense just to return an invalid statuscode? Same goes for read raw.